### PR TITLE
Added Pangolin Connector

### PR DESCRIPTION
**Pull request by prodesert22**
Pangolin is a Dex on Avalanche, similar to Uniswap

- [x] Review and approve connector @pradyuman-verma 
- [x] Deploy the contract and comment the address
- [x] Created resolver. found [here](https://github.com/Instadapp/dsa-resolvers/pull/42)
- [ ] Add to dsa-connect

### CODE REVIEW

**Comment by pradyuman-verma** [comment]
Hey @prodesert22, Can you rebase this branch with our latest commits?
As the repo has enabled Typescript support and many other changes, it will be easy for us to review it then!
Also, a lot of scripts that you have added are not needed now.

File: scripts/getMasterSigner_avalanche.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    method: "hardhat_impersonateAccount",
+    params: [masterAddress],
+  });
+  await wallet3.sendTransaction({
+    to: masterAddress,
+    value: ethers.utils.parseEther("10"),
+  });
+
+  return await ethers.getSigner(masterAddress);
+};
```
Comment (pradyuman-verma) [inline scripts/getMasterSigner_avalanche.js]: This script is not needed now, as we have updated the original script.

File: scripts/buildDSAv2_avalanche.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+const instaImplementations_m1 = require("../deployements/mainnet/Implementation_m1.sol/InstaImplementationM1.json")
+
+module.exports = async function (owner) {
+    const instaIndex = await ethers.getContractAt(abis.core.instaIndex, "0x6CE3e607C808b4f4C26B7F6aDAeB619e49CAbb25")
+
+    const tx = await instaIndex.build(owner, 2, owner);
+    const receipt = await tx.wait()
+    const event = receipt.events.find(a => a.event === "LogAccountCreated")
+    return await ethers.getContractAt(instaImplementations_m1.abi, event.args.account)
+};
```
Comment (pradyuman-verma) [inline scripts/buildDSAv2_avalanche.js]: This is not needed too.

**Comment by prodesert22** [comment]
I ended up adding the staking connector, stake LP tokens in MiniChefV2 and Single Stake PNG, the resolver it for the staking connector will be adding in the next days

File: test/pangolin/pangolin.test.js [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+const { web3, deployments, waffle, ethers } = hre;
+const { provider, deployContract } = waffle
+
+const abis = require("../../scripts/constant/abis");
+const buildDSAv2 = require("../../scripts/buildDSAv2_avalanche")
+const deployAndEnableConnector = require("../../scripts/deployAndEnableConnector.js")
+const encodeSpells = require("../../scripts/encodeSpells.js")
+const getMasterSigner = require("../../scripts/getMasterSigner_avalanche")
+
+const connectV2PangolinArtifacts = require("../../artifacts/contracts/avalanche/connectors/pangolin/main.sol/ConnectV2PngAvalanche.json");
```
Comment (pradyuman-verma) [inline test/pangolin/pangolin.test.js]:
```suggestion
import { ConnectV2PngAvalanche__factory } from "../../../typechain";
```

File: test/pangolin/pangolin.test.js [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+                    },
+                },
+            ],
+        });
+        masterSigner = await getMasterSigner(wallet3);
+        instaConnectorsV2 = await ethers.getContractAt(abis.core.connectorsV2, "0x127d8cD0E2b2E0366D522DeA53A787bfE9002C14");
+        // Deploy and enable Pangolin Connector
+        pangolinConnector = await deployAndEnableConnector({
+            connectorName: pangolinConnectorName,
+            contractArtifact: connectV2PangolinArtifacts,
```
Comment (pradyuman-verma) [inline test/pangolin/pangolin.test.js]:
```suggestion
            contractArtifact: ConnectV2PngAvalanche__factory,
```

File: test/avalanche/pangolin/pangolin_exchange.test.ts [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+const WAVAX_ADDRESS = "0xB31f66AA3C1e785363F0875A1B74E27b85FD66c7";
+const PNG_AVAX_LP_ADDRESS = "0xd7538cABBf8605BdE1f4901B47B8D42c61DE0367";
+
+describe("Pangolin DEX - Avalanche", function () {
+    const pangolinConnectorName = "PANGOLIN-TEST-A"
+    
+    let dsaWallet0: any;
+    let masterSigner: any;
+    let instaConnectorsV2: any;
+    let pangolinConnector: any;
```
Comment (pradyuman-verma) [inline test/avalanche/pangolin/pangolin_exchange.test.ts]:
```suggestion
    let dsaWallet0: Signer;
    let masterSigner: Signer;
    let instaConnectorsV2: Contract;
    let pangolinConnector: Contract;
```

File: test/avalanche/pangolin/pangolin_exchange.test.ts [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+            signer: masterSigner,
+            connectors: instaConnectorsV2
+        });
+        console.log("Pangolin Connector address: "+ pangolinConnector.address);
+    })
+
+    it("Should have contracts deployed.", async function () {
+        expect(!!instaConnectorsV2.address).to.be.true;
+        expect(!!pangolinConnector.address).to.be.true;
+        expect(!!masterSigner.address).to.be.true;
```
Comment (pradyuman-verma) [inline test/avalanche/pangolin/pangolin_exchange.test.ts]:
```suggestion
        expect(!!(await masterSigner.getAddress())).to.be.true;
```

File: test/avalanche/pangolin/pangolin_exchange.test.ts [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+      });
+    
+    describe("DSA wallet setup", function () {
+        it("Should build DSA v2", async function () {
+            dsaWallet0 = await buildDSAv2(wallet0.address)
```
Comment (pradyuman-verma) [inline test/avalanche/pangolin/pangolin_exchange.test.ts]: Replace everyplace where it is used!
```suggestion
            dsaWallet0 = await buildDSAv2(wallet0.getAddress())
```

File: test/pangolin/pangolin.test.js [hunk 8]
```diff
@@ -0,0 +1,204 @@
+const { expect } = require("chai");
+const hre = require("hardhat");
+
+const { web3, deployments, waffle, ethers } = hre;
+const { provider, deployContract } = waffle
```
Comment (pradyuman-verma) [inline test/pangolin/pangolin.test.js]:
```suggestion

```

File: test/pangolin/pangolin.test.js [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+describe("Pangolin DEX - Avalanche", function () {
+    const pangolinConnectorName = "PANGOLIN-TEST-A"
+    
+    let dsaWallet0
+    let masterSigner;
+    let instaConnectorsV2;
+    let pangolinConnector;
+    
+    const wallets = provider.getWallets()
+    const [wallet0, wallet1, wallet2, wallet3] = wallets
```
Comment (pradyuman-verma) [inline test/pangolin/pangolin.test.js]:
```suggestion
[wallet0, wallet1] = await ethers.getSigners();
```

File: test/pangolin/pangolin.test.js [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+const WAVAX_ADDRESS = "0xB31f66AA3C1e785363F0875A1B74E27b85FD66c7";
+const PNG_AVAX_LP_ADDRESS = "0xd7538cABBf8605BdE1f4901B47B8D42c61DE0367";
+
+describe("Pangolin DEX - Avalanche", function () {
+    const pangolinConnectorName = "PANGOLIN-TEST-A"
+    
+    let dsaWallet0
+    let masterSigner;
+    let instaConnectorsV2;
+    let pangolinConnector;
```
Comment (pradyuman-verma) [inline test/pangolin/pangolin.test.js]:
```suggestion
    let pangolinConnector: Contract;
  let wallet0: Signer, wallet1:Signer;
  let dsaWallet0: any;
  let instaConnectorsV2: Contract;
  let masterSigner: Signer;
```

File: test/pangolin/pangolin.test.js [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+            signer: masterSigner,
+            connectors: instaConnectorsV2
+        });
+        console.log("Pangolin Connector address: "+ pangolinConnector.address);
+    })
+
+    it("Should have contracts deployed.", async function () {
+        expect(!!instaConnectorsV2.address).to.be.true;
+        expect(!!pangolinConnector.address).to.be.true;
+        expect(!!masterSigner.address).to.be.true;
```
Comment (pradyuman-verma) [inline test/pangolin/pangolin.test.js]:
```suggestion
 expect(!!(await masterSigner.getAddress())).to.be.true;
```

File: test/pangolin/pangolin.test.js [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+      });
+    
+    describe("DSA wallet setup", function () {
+        it("Should build DSA v2", async function () {
+            dsaWallet0 = await buildDSAv2(wallet0.address)
```
Comment (pradyuman-verma) [inline test/pangolin/pangolin.test.js]:
```suggestion
           dsaWallet0 = await buildDSAv2(wallet0.getAddress());
```

Comment (pradyuman-verma) [reply test/pangolin/pangolin.test.js]: replace at other places too

File: test/pangolin/pangolin.test.js [hunk 13]
```diff
@@ -0,0 +1,204 @@
+const { expect } = require("chai");
```
Comment (pradyuman-verma) [inline test/pangolin/pangolin.test.js]: move `test/pangolin/pangolin.test.js` to `test/avalanche/pangolin/pangolin.test.js`

File: test/pangolin/pangolin.test.js [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+const hre = require("hardhat");
+
+const { web3, deployments, waffle, ethers } = hre;
+const { provider, deployContract } = waffle
+
+const abis = require("../../scripts/constant/abis");
+const buildDSAv2 = require("../../scripts/buildDSAv2_avalanche")
+const deployAndEnableConnector = require("../../scripts/deployAndEnableConnector.js")
+const encodeSpells = require("../../scripts/encodeSpells.js")
+const getMasterSigner = require("../../scripts/getMasterSigner_avalanche")
```
Comment (pradyuman-verma) [inline test/pangolin/pangolin.test.js]: update them as well.
Some things here are not needed as the repo has been updated to support multichain support !

File: test/avalanche/pangolin/pangolin_exchange.test.ts (hunk 5 above)
Comment (prodesert22) [reply test/avalanche/pangolin/pangolin_exchange.test.ts]: I tested and show an error message saying that dsaWallet0 is a contract and not a signer

File: test/pangolin/pangolin.test.js (hunk 3 above)
Comment (prodesert22) [reply test/pangolin/pangolin.test.js]: This file was moved to pangolin_exchange.test.ts, please ignore this file, It was a mistake of mine not erased it

File: test/avalanche/pangolin/pangolin_exchange.test.ts (hunk 5 above)
Comment (pradyuman-verma) [reply test/avalanche/pangolin/pangolin_exchange.test.ts]: Oh yes, it is indeed a contract.

File: contracts/avalanche/connectors/pangolin/exchange/main.sol [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+import { Events } from "./events.sol";
+
+abstract contract PangolinResolver is Helpers, Events {
+    /**
+     * @dev Deposit Liquidity.
+     * @notice Deposit Liquidity to a Pangolin pool.
+     * @param tokenA The address of token A.(For AVAX: 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)
+     * @param tokenB The address of token B.(For AVAX: 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)
+     * @param amtA The amount of A tokens to deposit.
+     * @param unitAmt The unit amount of of amtB/amtA with slippage.
```
Comment (pradyuman-verma) [inline contracts/avalanche/connectors/pangolin/exchange/main.sol]:
```suggestion
     * @param unitAmt The unit amount of amtB/amtA with slippage.
```

**Comment by pradyuman-verma** [comment]
@prodesert22 Can you add NatSpec comments on `main.sol` files?
 We do a check for this!

For example - https://github.com/Instadapp/dsa-connectors/blob/main/contracts/mainnet/connectors/COMP/main.sol

**Review by pradyuman-verma — APPROVED** [review]

**Review by pradyuman-verma — APPROVED** [review]

**Comment by pradyuman-verma** [comment]
@prodesert22 
Can you attach a screenshot of test cases, like they are passing or not!

Afterward, you can proceed towards deployment and comment down the address!

**Comment by pradyuman-verma** [comment]
@prodesert22 Also, some checks are failing.
Can you look into it?

**Comment by prodesert22** [comment]
Ok @pradyuman-verma, I'll check

**Comment by prodesert22** [comment]
> @prodesert22 Can you attach a screenshot of test cases, like they are passing or not!
> 
> Afterward, you can proceed towards deployment and comment down the address!

The log of tests are [here](https://pastebin.com/hyJCQ9cP)

**Comment by pradyuman-verma** [comment]
> Ok @pradyuman-verma, I'll check

OK sure

**Comment by prodesert22** [comment]
> @prodesert22 Also, some checks are failing. Can you look into it?

Hey @pradyuman-verma This is out of my reach, I believe you need to change the trigger for pull_request_target in[ status.yml](https://github.com/Instadapp/dsa-connectors/blob/main/.github/workflows/status.yml)
 Here contains a comment about this https://github.com/actions/labeler/issues/12#issueecomment-991806288

**Comment by pradyuman-verma** [comment]
> > @prodesert22 Also, some checks are failing. Can you look into it?
> 
> Hey @pradyuman-verma This is out of my reach, I believe you need to change the trigger for pull_request_target in[ status.yml](https://github.com/Instadapp/dsa-connectors/blob/main/.github/workflows/status.yml) Here contains a comment about this [actions/labeler#12 (comment)](https://github.com/actions/labeler/issues/12#issueecomment-991806288)

@ishanjain28 @thrilok209  Can you help here?

**Comment by ishanjain28** [comment]
Hey @prodesert22 Can you rebase the PR on `main`? I pushed a fix to make sure this check can run on PRs by external contributors.

**Comment by prodesert22** [comment]
Done @pradyuman-verma @ishanjain28

**Comment by pradyuman-verma** [comment]
> Done @pradyuman-verma @ishanjain28

Great!
Now you can deploy the contracts and comment down the addresses.

**Comment by prodesert22** [comment]
### Deployed on Avalanche

**Name:** `PANGOLIN-A`
**Addr:** [0x6798A4fe4420F08De26Ff3062492C057c4eB9a74](https://snowtrace.io/address/0x6798a4fe4420f08de26ff3062492c057c4eb9a74#code)
**tx:** [here](https://snowtrace.io/tx/0x4aab50d67a97fcd0ca6f4a66b136e6f2f69d020ef4a9118992ef1a562c3107f7)

**Name:** `PANGOLIN-STAKE-A`
**Addr:** [0x7f4D670Cd96b951E47Df3Bd9ee7948D457667117](https://snowtrace.io/address/0x7f4d670cd96b951e47df3bd9ee7948d457667117#code)
**tx:** [here](https://snowtrace.io/tx/0x4aab50d67a97fcd0ca6f4a66b136e6f2f69d020ef4a9118992ef1a562c3107f7)

**Review by thrilok209 — APPROVED** [review]

**Pull request merged**